openedx/edx-ora2
# feat: update `initialize` endpoint and create `assessments/feedback` endpoint in ORA Staff Grader

## Description
> This cover letter was based on [this PR](https://github.com/openedx/edx-ora2/pull/2095)

For this PR we have the following goals:

1. Create two new endpoints:
    - The `/api/ora_staff_grader/assessments/feedback/from` endpoint to list the assessment received for a specific submission of a learner.
    - The `/api/ora_staff_grader/assessments/feedback/to` endpoint to list the assessment given by a learner based on their submission ID.
2. Update the `/api/ora_staff_grader/initialize` endpoint to add the `fullname` and `email` fields.

> [!NOTE]  
> [This PR](https://github.com/openedx/frontend-app-ora-grading/pull/263) in the ORA Grading MFE will use these endpoints to display a more detailed table to the instructor.

## Dependecies

> [!IMPORTANT]  
> This upgrade will consist of two PRs: one for ORA and the other for edx-platform. This is because the main serializer of the service resides in the platform. To test it, you must fetch and set up the platform version with the serializer modification.
> - **edx-platform PR**: https://github.com/openedx/edx-platform/pull/33632

## What changed?

### 1. New endpoints

#### `/api/ora_staff_grader/assessments/feedback/from`

This endpoint fetches assessment details for a given:
- Submission UUID
- ORA Location

#### `/api/ora_staff_grader/assessments/feedback/to`

This endpoint fetches assessment details for a given:
- Submission UUID
- ORA Location

Both endpoints return the scorer information and scoring details to fill the following Staff Grade MFE table:

![image](https://github.com/openedx/edx-ora2/assets/64033729/f6792f0a-c902-48c1-9003-3e79685dc79d)

The following is an example of the response from either of the two endpoints:

```json
{
   "assessments": [
      {
         "assessment_id" : "8",
         "scorer_name" : "Nikola Tesla",
         "scorer_username" : "nikola",
         "scorer_email" : "nikola@tesla.com",
         "assesment_date" : "2024-02-15 15:29:55.374959+00:00",
         "assesment_scores" : [
            {
               "criterion_name" : "Ideas",
               "score_earned" : 3,
               "score_type" :" Fair"
            },
            {
               "criterion_name" :" Content",
               "score_earned" : 1,
               "score_type" :" Fair"
            }
         ],
         "problem_step" : "Peer",
         "feedback" : "Good Job"
      }
   ]
}
``` 

### 2. Updated endpoint

#### `/api/ora_staff_grader/initialize`

This is a pre-existing endpoint to which two extra fields are added (`fullname`, `email`) to fill the following table:

![image](https://github.com/openedx/edx-ora2/assets/64033729/61e419a1-ab54-476f-94a1-46282316465d)

The following is an example of the response from this endpoint:

```json
...
    "submissions" : {
        "619b8fb2-ebc0-49ed-9b35-4700efb28f66" : {
            "submissionUUID" : "619b8fb2-ebc0-49ed-9b35-4700efb28f66",
            "username" : "admin",
            "email" : "admin@admin.com",
            "fullname" : "Fernando Gonzalez",
            "teamName" : null,
            "dateSubmitted" : "2023-10-13 15:36:13.781179+00:00",
            "dateGraded" : "2023-10-17 20:26:07.146547+00:00",
            "gradedBy" : "admin",
            "gradeStatus" : "graded",
            "lockStatus" : "unlocked",
            "score" : {
                "pointsEarned" : 10,
                "pointsPossible" : 10
            }
        }
    },
``` 

## Developer Checklist


## Testing Instructions

To test this, you can use Postman:

1. Get an auth token from `/oauth2/access_token/` with the following **x-www-form-urlencoded** content type body:

    ```
    client_id: login-service-client-id
    client_secret: <your-client-secret>
    grant_type: password
    username: <your-username>
    password: <your-password>
    ```
    From the response, copy the `access_token`, which must be sent in the headers of the endpoints.

3. Test the endpoints:
- `/api/ora_staff_grader/initialize`
    - **Query Params**: oraLocation
- `/api/ora_staff_grader/assessments/feedback/from`
    - **Query Params**: oraLocation, submissionUUID
- `/api/ora_staff_grader/assessments/feedback/to`
    - **Query Params**: oraLocation, submissionUUID
 
**NOTE**: From the MFE ora-grading you can get the `oraLocation` from here:

![image](https://github.com/openedx/edx-ora2/assets/64033729/fcd10b8c-cbbd-427e-93de-334f8c0ed7e1)

And the `submissionUUID` from the response of the `initialize` endpoint here:

![image](https://github.com/openedx/edx-ora2/assets/64033729/340943e3-f735-4009-9e24-b52e84ac625c)

Make sure that the `oraLocation` value is encoded.

![image](https://github.com/openedx/edx-ora2/assets/64033729/9a2f01f7-96bc-4856-b08c-8d353eb88bb0)

## Reviewer Checklist

Collectively, these should be completed by reviewers of this PR:


FYI: @openedx/content-aurora
